pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SnapshotSegmentAbortedTxnProcessorImpl.java

```diff
@@ -0,0 +1,758 @@
+            UpdateIndex,
+            WriteSegment,
+            DeleteSegment,
+            Clear
+        }
+
+        private volatile OperationState operationState = OperationState.None;
+
+        ConcurrentLinkedDeque<Pair<OperationType, Pair<CompletableFuture<Void>,
+                Supplier<CompletableFuture<Void>>>>> taskQueue = new ConcurrentLinkedDeque<>();
+
+        public PersistentWorker(PersistentTopic topic) {
+            this.topic = topic;
+            this.snapshotSegmentsWriterFuture =  this.topic.getBrokerService().getPulsar()
+                    .getTransactionBufferSnapshotServiceFactory()
+                    .getTxnBufferSnapshotSegmentService().createWriter(TopicName.get(topic.getName()))
+                    .exceptionally(ex -> {
+                        log.error("{} Failed to create snapshot index writer", topic.getName());
+                        topic.close();
+                        return null;
+                    });
+            this.snapshotIndexWriterFuture =  this.topic.getBrokerService().getPulsar()
+                    .getTransactionBufferSnapshotServiceFactory()
+                    .getTxnBufferSnapshotIndexService().createWriter(TopicName.get(topic.getName()))
+                    .exceptionally((ex) -> {
+                        log.error("{} Failed to create snapshot writer", topic.getName());
+                        topic.close();
+                        return null;
+                    });
+        }
+
+        public CompletableFuture<Void> appendTask(OperationType operationType,
+                                                  Supplier<CompletableFuture<Void>> task) {
+            CompletableFuture<Void> taskExecutedResult = new CompletableFuture<>();
+            switch (operationType) {
+                case UpdateIndex -> {
+                    /*
+                      The update index operation can be canceled when the task queue is not empty,
+                      so it should be executed immediately instead of appending to the task queue.
+                      If the taskQueue is not empty, the worker will execute the tasks in the queue.
+                     */
+                    if (!taskQueue.isEmpty()) {
+                        executeTask();
+                        return CompletableFuture.completedFuture(null);
+                    } else if (STATE_UPDATER.compareAndSet(this, OperationState.None, OperationState.Operating)) {
+                        return task.get().whenComplete((ignore, throwable) -> {
+                            if (throwable != null && log.isDebugEnabled()) {
+                                log.debug("[{}] Failed to update index snapshot", topic.getName(), throwable);
+                            }
+                            STATE_UPDATER.compareAndSet(this, OperationState.Operating, OperationState.None);
+                        });
+                    } else {
+                        return CompletableFuture.completedFuture(null);
```

Comment:
The operation of updating indexes has yet to be complete when this future is complete, right?

Reply:
If there is another operation executing, this operation can be canceled. It is just a timer worker. Completing it is enough.